platform/engineering/request-for-change/2019-10-16-terraform-automation-implementation-RFC.md

```diff
@@ -0,0 +1,43 @@
+### Placeholder, will work on this more tomorrow ### 
+
+# RFC: _Necessary Implementation of Automation for Terraform Workflow(s)_
+
+- Date: _2019-10-16_
+- Related Issue: _git/terraform issue(s)_
+
+## Background
+_Explain the current state of the feature._
+
+Currently, we are running terraform (plan and apply) from workstations manually. When a PR is opened, the expectation is that they add a `plan` output, it's approved, the change is `apply`'ed, then merged into master.
+
+## Motivation
+_Why do we want to change the current implementation? What problem(s) does the change solve?_
+
+The workflow described above worked for a while when there were only a 2-3 engineers working really closely together. While not ideal, there wasn't a strong enough need to add additional systems for a while. With more engineers onboarded and making changes, we see a number of problems:
+
+1. Requires a workstation have write access to prod which:
+   - imits the pool of folks able to contribute
+   - weakens security posture
+1. Involves manual steps of uploading a plan, then doing the apply/merge in the correct order
+
+Wrapping this workflow in some for of automation will improve our process in multiple ways. It will help increase new hire velocity enabling them to `terraform plan` even before they get direct access to AWS. It could increase overall velocity by automating a currently manual, repetitive step. Finally, it will also provide us with a means to add a step for `terraform fmt` which will keep our terraform files clean and standardized. 
+
+Additionally, there are times where edits to infrastructure may happen outside of Terraform (for example: someone is testing a setting change in an ELB by tweaking it in dev then forgets to commit back), and when it's time to run a `plan` there's unclean state that has to be addressed first.
+
```

Comment:
A thing that's missing so far is having modules in a single repo without versioning. I'd also like to address this. Could be this one or another RFC if we think that's too much to take on at once.

Reply:
I personally think that would be a separate issue from the implementation of automation for tf fmt, plan, apply. However, I would also be interested to hear others' opinion on it.

Reply:
I think it could be it's own RFC. I would be interesting in working on that module piece as well. Let me know how I can get involved please!